spec/unit/shoulda/matchers/active_record/have_rich_text_matcher_spec.rb

```diff
@@ -0,0 +1,99 @@
+require 'unit_spec_helper'
+
+describe Shoulda::Matchers::ActiveRecord::HaveRichText, type: :model do
+  shared_examples 'undefined rich text' do |kind|
+    it 'raise a undefined method error when action text is not defined' do
+      create_table_posts
+  
+      invalid_record = -> { new_post(has_rich_text_association: true) }
+
+      message = /undefined method `has_rich_text' for Post/
+        
+      expect(&invalid_record).to raise_error(
+        NoMethodError,
+        message
+      )
+    end
+
+    include_examples "invalid attribute", kind
+  end
+
+  shared_examples "invalid attribute" do |kind|
+    it "does not matches when the provided attribute it's not a RichText class" do
+      create_table_posts(has_content_attribute: true)
+  
+      invalid_model = define_model_class(:Post)
+      invalid_record = invalid_model.new(content: "Content")
+      expected_message = "Expected Post to have :content as a " \
+        "ActionText::RichText but :content is a #{kind}"
+  
+      expect { have_rich_text(:content) }.
+        not_to match_against(invalid_record).
+        and_fail_with(expected_message)
+    end
+  end
+
+  def self.rich_text_is_defined?
+    defined?(ActionText::RichText)
+  end
+
+  if rich_text_is_defined?
+    it 'returns the message including the name of the provided attribute' do
+      matcher = have_rich_text(:content)
+      expect(matcher.description).
+        to eq('have :content as a ActionText::RichText')
+    end
+
+    context 'a has_rich_text association' do
+      include_examples "invalid attribute", 'String'
+
+      it 'matches when the subject configures has_rich_text for a custom attribute' do
+        create_table_posts
+    
+        valid_record = new_post(has_rich_text_association: true)
+    
+        expect(valid_record).to have_rich_text(:content)
+      end
+    
+      it 'does not matches when lacking has_rich_text attribute' do
+        create_table_posts
+    
+        invalid_record = new_post
+        expected_message = 'Expected Post to have :invalid_attribute as a ' \
+          'ActionText::RichText but :invalid_attribute does not exist'
+        
+        expect { have_rich_text(:invalid_attribute) }.
+          not_to match_against(invalid_record).
+          and_fail_with(expected_message)
+      end
+
+      it 'fails with the appropriate message when negated' do
+        valid_record = new_post(has_rich_text_association: true)
+        expected_message = 'Did not expect Post to have a rich text :content'
+
+        expect { expect(valid_record).not_to have_rich_text(:content) }.
+          to fail_with(expected_message)
+      end
+    end
+  elsif active_record_version >= 4 && active_record_version < 5
+    include_examples 'undefined rich text', 'NilClass'
+  else
+    include_examples 'undefined rich text', 'String'
+  end
+
+  def create_table_posts(has_content_attribute: false)
```

Comment:
Can you combine these two methods together? `define_model` calls `create_table` under the hood.

For instance:

``` ruby
def new_post(has_content_attribute: false, is_rich_text_attribute: false)
  columns = {}

  if has_content_attribute
    columns[:content] = :string
  end

  define_model 'Post', columns do
    if has_content_attribute && is_rich_text_attribute
      has_rich_text :content
    end
  end.new
end
```

Also, note that I've renamed the `has_rich_text_association`. We need to be careful when using the word "association". The `has_rich_text` macro does _not_ make `content` an association, so I feel like that could be confusing if we use that word.

Reply:
Hi, @mcmire, thank you for your support!

If i’m not wrong has_rich_text macro create a association between the model and the ActionTextRichText as a has_one. e.g.: Post has one ActionText through content.

The problem here was some silly mistakes made by me.

One was the poor description of the tests caused by my rusty English, I need to work on it.

The second was because I used content for the relationship with RichText and used the same name for a attribute. It is easy to think that they are the same...

The content attribute is used for failures tests only. It's for a scenario when the developer try to use a attribute of the model and the matcher raise a warning informing something like: this is not valid, it was expecting a ActionText, but this is a Integer/String/Etc.

I refactored the tests and add your suggestion.

What do you think?

Reply:
You're totally right about the association, I should have checked that. That's what you get for making assumptions, I guess :/ So thanks for clarifying that. I checked over your changes again!